Introduce String predicates as validators

dvberkel opened the pull request:
This pull request introduces a number of String predicates as validators. Among them are `isBlank`, `isEmpty`, `contains` etcetera.

The only predicates that are missing as validators are

* `contains` that accepts a regular expression.
* `hasSurrogatePairAt`
* `matches`
* `regionMatches`

The order of the validators and their corresponding tests is changed. They are now in alphabetical order.

Furthermore the `allDigit` validator is rewritten with the use of the now available `any` validator and the `Char::isDigit` predicate.

dvberkel committed:
order current String validators alphabetically

dvberkel committed:
Implement String predicates as validators

The following three predicates aren't implemented
* `contains` which accepts a regex
* `hasSurrogatePairAt`
* `regionMatches`

geertmulders commented on konfork-core/src/commonMain/kotlin/io/github/konfork/core/validators/Strings.kt:
```diff
@@ -6,70 +6,101 @@ import io.github.konfork.core.Specification
-fun <C, E> Specification<C, String, E>.pattern(hint: HintBuilder<C, String, E>, pattern: Regex) =
-    addConstraint(hint, pattern) { it.matches(pattern) }
+fun <C> Specification<C, String, String>.allDigits() =
+    all(stringHint("is not all digits"), Char::isDigit)
```
It would be nicer to call the other `allDigits` method here.

dvberkel replied:
:+1: 

I wasn't thinking here

geertmulders commented on konfork-core/src/commonMain/kotlin/io/github/konfork/core/validators/Strings.kt:
```diff
@@ -6,70 +6,101 @@ import io.github.konfork.core.Specification
+    char: Char,
+    ignoreCase: Boolean = false
+) =
+    addConstraint(hint) { it.contains(char, ignoreCase) }
```
What I'm missing here (and some other places) are the `char` and `ignoreCase` parameters in the `addConstraint` call: 

```kotlin
addConstraint(hint, char, ignoreCase) { it.contains(char, ignoreCase) }
```

Adding these will make them available in the `HintBuilder`, which is nice if someone wishes to make its own hints.

dvberkel replied:
Will apply this to the places it makes sense

geertmulders commented on konfork-core/src/commonMain/kotlin/io/github/konfork/core/validators/Strings.kt:
```diff
@@ -6,70 +6,101 @@ import io.github.konfork.core.Specification
-    return addConstraint(hint, range.first, range.last) { it.length in range }
-}
+fun <C> Specification<C, String, String>.any(predicate: (Char) -> Boolean) =
+    any(stringHint("none of the characters comply"), predicate)
```
In many places it might be better to use `staticHint` instead of `stringHint`

dvberkel replied:
:+1:

geertmulders commented on konfork-core/src/commonMain/kotlin/io/github/konfork/core/validators/Strings.kt:
```diff
@@ -6,70 +6,101 @@ import io.github.konfork.core.Specification
+    } else {
+        ""
+    }
+    contains(stringHint("does not contain character '$char'$message"), char, ignoreCase)
```
We have to make sure to return a `ConstrainBuilder` from all validator-builders.

dvberkel replied:
This makes sense

dvberkel committed:
Move isAllDigits to CheckDigits file

It wasn't used in the String validators any more

Fixes #7

dvberkel committed:
reuse the allDigits validator

resolve feedback https://github.com/konfork/konfork/pull/5#discussion_r1035631788

dvberkel committed:
use staticHint in String validators

resolves feedback https://github.com/konfork/konfork/pull/5#discussion_r1035639266

dvberkel committed:
String validators always return ConstraintBuilder

resolves feedback https://github.com/konfork/konfork/pull/5#discussion_r1035685944

dvberkel committed:
Provide values to String validator HintBuilder

resolves feedback https://github.com/konfork/konfork/pull/5#discussion_r1035634632

dvberkel committed:
Replace pattern validator with matches and contains validators

Fixes #8

dvberkel committed:
Remove isAllDigits test

The isAllDigits isn't used for a validator and therefore was moved to the CheckDigits predicates and made private. Therefore the test is superfluous

gmulders commented:
Looks great @dvberkel , do you merge the PR or shall I?
